core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -544,3 +645,81 @@ private[storage] class PartiallyUnrolledIterator[T](
     iter = null
   }
 }
+
+/**
+ * A wrapper which allows an open [[OutputStream]] to be redirected to a different sink.
+ */
+private class RedirectableOutputStream extends OutputStream {
+  private[this] var os: OutputStream = _
+  def setOutputStream(s: OutputStream): Unit = { os = s }
```

Comment:
Do we need a whole new class for this? Can we just write (failed stream + the rest of input stream) to the file when we drop it to disk?